scripts/circleci_scraper/scraper.py

```diff
@@ -0,0 +1,348 @@
+            for job in jobs.items:
+                if job.name in job_names:
+                    self.export_test_metadata_by_job(organization, repository, workflow.name, job)
+                    self.export_test_artifacts_by_job(organization, repository, workflow.name, job)
+            next_page_token = jobs.next_page_token
+            if not next_page_token:
+                break
+
+    def export_test_metadata_by_job(
+        self,
+        organization: str,
+        repository: str,
+        workflow_name: str,
+        job: Job,
+    ) -> None:
+        """Export test metadata for a specific job.
+
+        Args:
+            organization (str): The organization name.
+            repository (str): The repository name.
+            workflow_name (str): The workflow name.
+            job (Job): The job details.
+
+        Raises:
+            CircleCIClientError: If there is an error in the CircleCI API request.
+            CircleCIScraperError: If there is an error in downloading the artifacts.
+        """
+        file_content: dict[str, Any] = {"job": job.dict(), "test_metadata": []}
+        next_page_token: str | None = None
+        while True:
+            test_metadata: TestMetadataGroup = self._client.get_test_metadata(
+                organization, repository, str(job.job_number), next_page_token
+            )
+            if not test_metadata:
+                break
+            file_content["test_metadata"] += [item.dict() for item in test_metadata.items]
+            next_page_token = test_metadata.next_page_token
+            if not next_page_token:
+                break
+        if not file_content["test_metadata"]:
+            self.logger.info(f"There is no test metadata for job {str(job.job_number)}")
+            return
+        self.export_test_metadata(repository, workflow_name, job, file_content)
+
+    def export_test_metadata(
+        self, repository: str, workflow_name: str, job: Job, file_content: dict[str, Any]
+    ):
+        """Export test metadata.
+
+        Args:
+            repository (str): The repository name.
+            workflow_name (str): The workflow name.
+            job (Job): The job details.
+            file_content (Dict[str, Any]): The content to be output.
+
+        Raises:
+            CircleCIClientError: If there is an error in the CircleCI API request.
+            CircleCIScraperError: If there is an error in downloading the artifacts.
+        """
+        test_metadata_directory: str = os.path.join(
+            self._test_result_dir, repository, workflow_name, job.name, self._test_metadata_dir
+        )
+        if not os.path.isdir(test_metadata_directory):
+            os.makedirs(test_metadata_directory)
+        file_path: str = os.path.join(test_metadata_directory, f"{job.job_number}.json")
+        if not os.path.exists(file_path):
+            with open(file_path, "w") as file:
+                file.write(json.dumps(file_content, default=str))
```

Comment:
As a non-Python person, does this auto-close the file handles, or do we need to explicitly close those?

Reply:
Python uses indentation to indicate blocks, like how Javascript wraps them with { }. 
"with" is declarative, so yes, at the end of the "block" the handle is marked for disposal.